accepted/0000-link-packages-to-source-and-build.md

```diff
@@ -0,0 +1,514 @@
+- Linking packages to the source and build.
+- Doesn't expose any personal identifiable information about maintainers.
+- Avoids developer-managed keys by placing trust in the CI/CD identity provider.
+
+Today this means only commercial CI/CD providers will be supported. This is far from where all developers are publishing today so this will add friction and slow adoption. All major CI/CD providers offer free plans for Open Source projects and we hope that all packages will eventually be built out in the open to make supply chain attacks harder to execute.
+
+### Threat model
+Our threat model focuses on package hijacking attacks: where a developer with good intentions, writing secure code (i.e. no accidental vulnerabilities) publishes this package to the npm registry.
+
+The attacker might then gain access to the maintainers login credentials and upload a modified version of the package to the registry under a different, in-range version number causing end-developers who depend on this package to consume malicious code. This maliciously crafted package is often not reflected in the package's source code repository and uploaded directly from the attacker's machine.
+
+This proposal does not mitigate against compromised npm login credentials. The aim is to make it harder to execute these types of attacks by creating a public audit trail for where, how and who published a package. Over time the presence of this information can be enforced.
+
+As such, what we care about in this threat model is whether this provenance information can be forged and whether verification can be bypassed.
+
+The following table highlights the impact of different compromises:
+
+| Compromised infrastructure | Forge provenance                 | Bypass verification |
+|----------------------------|----------------------------------|--------------------|
+| npm registry               | No                               | Yes                |
+| CI/CD                      | Yes                              | No                 |
+| Sigstore                   | Yes                              | No                 |
+| Network (MITM)             | No                               | Yes                |
+| User account (npm)         | No                               | Yes                |
+| User account (GH)          | No                               | No                 |
+| Trusted builder            | Likely                           | No                 |
+
+An attacker could forge what goes in the provenance information if the CI/CD provider or Sigstore was compromised. Similarly, if the trusted builder was compromised in some way it's likely the attacker could forge the provenance information during the build. Mitigating these attacks is out of scope in this proposal.
+
+An attacker could bypass verification if the npm registry, npm user or the network was compromised by not providing provenance information or removing it in transit.
```

Comment:
> Does this imply that `npm install` would fail open if no provenance info is provided?

Yes I think we need to go with fail open by default until some large portion of packages have provenance.

> Could an option be added to require it?

Yes! I'd love to provide ways for consumers to require provenance and specify policies around which packages or org scopes for example are required 👍  Outlined a similar idea under "Sigstore integration in the npm CLI > Verifying"

Reply:
It's probably worth reviewing/discussing this in the context of https://github.com/npm/rfcs/pull/182 which has been pending the now-shipped `npm query` that also has plans for policies that are related to audit :)

Reply:
requiring provenance should be sticky, if a package stops providing provenence data it should fail the install and require intervention.

There should be similar warnings for publishers.

Reply:
> requiring provenance should be sticky, if a package stops providing provenence data it should fail the install and require intervention.

Yes I would love to enforce this somehow. There probably needs to be an escape hatch if you need to move publishing to some platform that isn't supported for example, but should require some hoops to disable, e.g. 2FA must be used to allow publishing a new version without provenance information.

Reply:
note: This is the "SLSA policy" question, which is a bit complicated. See my other comment below (line 365). It might be best to leave this to a follow-up RFC.